xdp-dump/xdpdump.c

```diff
@@ -756,21 +692,25 @@ static bool list_interfaces(struct dumpopt *cfg)
 	printf("--------  ----------------  "
 	       "--------------------------------------------------\n");
 	for (idx = indexes; idx->if_index; idx++) {
-		enum xdp_attach_mode mode;
-		struct bpf_prog_info info = {};
-		struct iface iface = {
-			.ifindex = idx->if_index,
-			.ifname = idx->if_name,
-		};
-
-		if (get_loaded_program(&iface, &mode, &info) != 0) {
+		struct xdp_multiprog *mp;
+
+		mp = xdp_multiprog__get_from_ifindex(idx->if_index);
+		if (IS_ERR_OR_NULL(mp)) {
 			printf("%-8d  %-16.16s  %s\n",
-			       iface.ifindex, iface.ifname,
+			       idx->if_index, idx->if_name,
 			       "<No XDP program loaded!>");
 		} else {
+			struct xdp_program *prog = NULL;
+
 			printf("%-8d  %-16.16s  %s()\n",
-			       iface.ifindex, iface.ifname,
-			       info.name);
+			       idx->if_index, idx->if_name,
+			       xdp_program__name(xdp_multiprog__dispatcher(mp)));
+
+			while ((prog = xdp_multiprog__next_prog(prog, mp)))
+				printf("%-25s %s()\n", "=>",
```

Comment:
The main reason for adding the guess_full_main_function() function was to solve the long program name issue, issue #17. But all of this is undone :) Even this command is not showing the long program loaded:

```
[root@xdp-dev xdp-dump]# ./xdpdump -D
if_index  if_name           XDP program entry function
--------  ----------------  --------------------------------------------------
1         lo                <No XDP program loaded!>
2         eth0              <No XDP program loaded!>
3         eth1              <No XDP program loaded!>
```

Old code shows this:

```
# ./xdpdump -D
if_index  if_name           XDP program entry function
--------  ----------------  --------------------------------------------------
1         lo                <No XDP program loaded!>
2         eth0              xdp_test_prog_w()
3         eth1              <No XDP program loaded!>
```

And the real long name is:

```
SEC("xdp_test_prog_long")
int xdp_test_prog_with_a_long_name(struct xdp_md *ctx)
{
    bpf_debug("PASS[1]: prog %u\n", ctx->ingress_ifindex);
    return XDP_PASS;
}
```

Also do not forget to take care of the case where there is more than one program matching, see the other error you removed in find_target_program() above.

I've attached a simeple ebpf program below that can be used two simulate the use cases:
[test_long_func_name.txt](https://github.com/xdp-project/xdp-tools/files/4585973/test_long_func_name.txt)

Reply:
Yes, I do realise I undid all your careful matching :)
As it says in the commit log, though, it's not completely accidental: I decided to sidestep the whole issue instead by just having libxdp enforce a max length. This is not ideal, of course, but I think it's something we ought to solve at the libxdp level.

My current idea is that since libxdp will pin all component programs anyway, we can put more information into the pin filenames. Such as the BTF ID of the function, which libxdp can then read back out when generating a struct xdp_program from a kernel fd, so we can make sure we always have the full name in xdp_program__name(), and we won't have to guess.

As for the other issue with more than one program name matching, that is actually worse than what you say: With XDP multiprogs we can have more than one sub-program using the same function name, since they are completely independent BPF objects. This means that program name is not guaranteed to be a valid disambiguator between functions; so I think xdpdump needs to switch to using IDs instead...

Reply:
> Yes, I do realise I undid all your careful matching :)
> As it says in the commit log, though, it's not completely accidental: I decided to sidestep the whole issue instead by just having libxdp enforce a max length. This is not ideal, of course, but I think it's something we ought to solve at the libxdp level.

Well, this is a problem, as you assume we will always have multi prog. The issues I raise are all in the non-multiprog setup. I.e. someone just loading a program. Which xdpdump must also support.

> 
> My current idea is that since libxdp will pin all component programs anyway, we can put more information into the pin filenames. Such as the BTF ID of the function, which libxdp can then read back out when generating a struct xdp_program from a kernel fd, so we can make sure we always have the full name in xdp_program__name(), and we won't have to guess.

Same as above, this is only in the non-mulitprog case.
 
> As for the other issue with more than one program name matching, that is actually worse than what you say: With XDP multiprogs we can have more than one sub-program using the same function name, since they are completely independent BPF objects. This means that program name is not guaranteed to be a valid disambiguator between functions; so I think xdpdump needs to switch to using IDs instead...

My idea was still to do our best guessing, i.e. match first xchars on program entries on various programs, and then walk that specific BPF tables. Because we use an exact match this is still fine. The only problem is when multiple programs use the same first x chars, this we can probably solve by allowing a program NR/index instead of a name in the --program-names option. So something like `--program-names=1,5` to trace the first and 5th program.


I think for now, as xdpdump does not yet support multi-programs, leave the code as is, and I fix it up when we add multi-program support as this area needs to change anyway. I think your patchset should still work with the old code.